In [64]:
import pandas as pd
from sklearn import linear_model
import numpy as np

In [18]:
data = pd.read_html('https://www.w3schools.com/python/python_ml_multiple_regression.asp')

In [49]:
df = data[1].loc[0:]
df = df.rename({0:'Car', 1:'Model', 2:'Volume', 3:'Weight', 4:'CO2'}, axis=1)

In [50]:
df.head()

Car       Model  Volume  Weight  CO2
0      Toyota        Aygo    1000     790   99
1  Mitsubishi  Space Star    1200    1160   95
2       Skoda      Citigo    1000     929   95
3        Fiat         500     900     865   90
4        Mini      Cooper    1500    1140  105

In [53]:
df.to_csv('data/CO2_emission.csv', index=False)

In [65]:
X = np.array(df[['Weight', 'Volume']])
y = np.array(df[['CO2']])

In [66]:
regr = linear_model.LinearRegression()
regr.fit(X, y)

LinearRegression()

In [67]:
#predict the CO2 emission of a car where the weight is 2300kg, and the volume is 1300cm3:
predictedCO2 = regr.predict([[2300, 1300]])

In [81]:
predictedCO2[0,0].round(2)

107.21

We have predicted that a car with 1.3 liter engine, and a weight of 2300 kg, will release approximately 107 grams of CO2 for every kilometer it drives.

## Coefficient
The coefficient is a factor that describes the relationship with an unknown variable.

Example: if x is a variable, then 2x is x two times. x is the unknown variable, and the number 2 is the coefficient.

In this case, we can ask for the coefficient value of weight against CO2, and for volume against CO2. 
The answer(s) we get tells us what would happen if we increase, or decrease, one of the independent values.

In [82]:
# Print the coefficient values of the regression object:
regr.coef_

array([[0.00755095, 0.00780526]])

## Result Explained
The result array represents the coefficient values of weight and volume.

Weight: 0.00755095

Volume: 0.00780526

These values tell us that if the weight increase by 1kg, the CO2 emission increases by 0.00755095g.

And if the engine size (Volume) increases by 1 cm3, the CO2 emission increases by 0.00780526 g.

I think that is a fair guess, but let test it!

We have already predicted that if a car with a 1300cm3 engine weighs 2300kg, the CO2 emission will be approximately 107g.

What if we increase the weight with 1000kg?

In [84]:
# Copy the example from before, but change the weight from 2300 to 3300:
predictedCO2 = regr.predict([[3300, 1300]])
predictedCO2[0,0]

114.75968006922292

We have predicted that a car with 1.3 liter engine, and a weight of 3300 kg, 

will release approximately 115 grams of CO2 for every kilometer it drives.

Which shows that the coefficient of 0.00755095 is correct:

In [86]:
107.2087328 + (1000 * 0.00755095)

114.75968280000001

In [97]:
regr.predict([[2300, 2300]]) # increase the volume from 1300 to 2300

array([[115.01399033]])

In [95]:
107.2087328 + (1000 * 0.00780526)

115.01399280000001

In [98]:
regr.predict([[1300, 1300]]) # decrease the weight from 2300 to 1300

array([[99.65778553]])

In [99]:
107.2087328 - (1000 * 0.00755095)

99.6577828